ibridges/search.py

```diff
@@ -47,90 +60,146 @@ def search_data(
-    data_name_query = session.irods_session.query(
-        icat.COLL_NAME, icat.DATA_NAME, icat.DATA_CHECKSUM
-    )
+    data_name_query = session.irods_session.query(icat.COLL_NAME, icat.DATA_NAME, icat.DATA_CHECKSUM)
+
     # iRODS queries do not know the 'or' operator, so we need three searches
     # One for the collection, and two for the data
     # one data search in case path is a collection path and we want to retrieve all data there
     # one in case the path is or ends with a file name
-    path_params = _path_params(path)
-    if path_params:
-        path, name, parent = path_params
-        # all collections starting with path
-        coll_query = coll_query.filter(icat.LIKE(icat.COLL_NAME, path))
-
-        # all data objects in path
-        data_query = data_query.filter(icat.LIKE(icat.COLL_NAME, path))
-        # all data objects on path.parent with name
-        data_name_query = data_name_query.filter(icat.LIKE(icat.DATA_NAME, name)).filter(
-            icat.LIKE(icat.COLL_NAME, parent)
-        )
-    if key_vals:
-        for key in key_vals:
-            data_query.filter(icat.LIKE(icat.META_DATA_ATTR_NAME, key))
-            coll_query.filter(icat.LIKE(icat.META_COLL_ATTR_NAME, key))
-            data_name_query.filter(icat.LIKE(icat.META_DATA_ATTR_NAME, key))
-            if key_vals[key]:
-                data_query.filter(icat.LIKE(icat.META_DATA_ATTR_VALUE, key_vals[key]))
-                coll_query.filter(icat.LIKE(icat.META_COLL_ATTR_VALUE, key_vals[key]))
-                data_name_query = data_name_query.filter(
-                    icat.LIKE(icat.META_DATA_ATTR_VALUE, key_vals[key])
-                )
+    if path is None:
+        path = session.home
+    path = IrodsPath(session, path)
+
+    # Filter only paths according to the search path.
+    coll_query = coll_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}/%"))
+    data_query = data_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}/%"))
+    data_name_query.filter(icat.LIKE(icat.COLL_NAME, f"{path}"))
+    queries = [coll_query, data_query, data_name_query]
+
+    if path_pattern is not None:
+        _path_filter(path, path_pattern, *queries)
+
+    if key is not ...:
+        if isinstance(key, str):
+            _key_filter(key, *queries)
+        else:
+            for sub_key in key:
+                _key_filter(sub_key, *queries)
+
+    if value is not ...:
+        if isinstance(value, str):
+            _value_filter(value, *queries)
+        else:
+            for sub_val in value:
+                _value_filter(sub_val, *queries)
+
+    if units is not ...:
+        if isinstance(units, str):
+            _units_filter(units, *queries)
+        else:
+            for sub_units in units:
+                _units_filter(sub_units, *queries)
 
     results = []
-    if checksum:
+    if checksum is not None:
+        if not checksum.startswith("sha2:"):
```

Comment:
I understand that you want to free the user of prefixing the checksum with sha. However iRODS servers know two checksums, sha2 and md5. Only the sha2 is prefixed with sha2 while md5 is the real md5 checksum without any additional prefix.

So in iRODS instances where md5 checksums are configured the search will fail to retrieve any results. :(